### ISSUE TITLE
separate include paths for each installed dependency

### ISSUE THREAD

**Issue by MarkSchuckParadigm** [microsoft/vcpkg#12112]
****

**Is your feature request related to a problem? Please describe.**

Before going further I want to make it clear, this is *my* fault but I think vcpkg could arrange the includes differently to prevent accidentally running into this problem. 

The issue is that vcpkg puts all dependency include files in a single root path so as soon as my library depends on a vcpkg managed dependency it gains access to all the other paths as well. In other words i can include headers that i haven't done a 'target_link_library(...)' on. this causes two main categories of issues:
1) 'multiply defined symbols'. if the header was part of a dependency that has dual use as header only or statically linked library (like spdlog) then I can have a static library (say, libA) that accidentally includes spdlog/spdlog.h without a 'target_link_libraries(<target> spdlog::spdlog)'. This library builds fine, but libA will have its own copy of the definitions for all the spdlog symbols and leads to errors for multiply defined symbols when linking to the final executable which was properly linking to spdlog.lib via target_link_libraries(app spdlog::spdlog). 
2)  'Can't find header xyz.h'. If the header was part of a always header only library (like many boost libraries) then everything works fine until i try compile my library in a system that is not using vcpkg. for example, using yocto/bitbake that uses unique paths per dependency / per target i will get include errors.

I also worry about ODR violations where a header changes its interface depending on compiler flags/definitions but they never get set because cmake doesn't know about the dependency. but I don't think that i've run into any so far

**Proposed solution**
all dependencies should get their own include folder.

Current:
<vcpkg_root>\installed\<triplet>\include\boost\...
<vcpkg_root>\installed\<triplet>\include\spdlog\spdlog.h
<vcpkg_root>\installed\<triplet>\include\spdlog\...
<vcpkg_root>\installed\<triplet>\include\fmt\fmt.h
<vcpkg_root>\installed\<triplet>\include\fmt\...

Proposed:
<vcpkg_root>\installed\<triplet>\boost\include\boost\...
<vcpkg_root>\installed\<triplet>\spdlog\include\spdlog\spdlog.h
<vcpkg_root>\installed\<triplet>\spdlog\include\spdlog\...
<vcpkg_root>\installed\<triplet>\fmt\include\fmt\fmt.h
<vcpkg_root>\installed\<triplet>\fmt\include\fmt\...
<vcpkg_root>\installed\<triplet>\zlib\include\zlib.h
<vcpkg_root>\installed\<triplet>\<...>\include\...

then the include path for each target would be unique. so for target_link_libraries(libA spdlog::spdlog) would then add '<vcpkg_root>\installed\<triplet>\spdlog\include\' to libA's include paths and would prevent it from accidentally including zlib.h.

**Describe alternatives you've considered**
AFAICS this is the only solution that fixes the problem without intrusively modifying the other projects (ex: having a different include for spdlog as header only vs. statically linked) or adding compiler defines for each project and check they match at the beginning of each include file.

**Additional context**
I took a quick look at existing open & closed issues but couldn't find any

**Comment by Neumann-A** [comment]
And what happens to cmake find modules? They won't find their INCLUDE_DIRs anymore since the don't expect another sub directory.

**Comment by gix** [comment]
> And what happens to cmake find modules? They won't find their INCLUDE_DIRs anymore since the don't expect another sub directory.

IMHO (vcpkg-built) libraries should only ever be included using `find_package(<name> CONFIG REQUIRED)`. The corresponding `<name>-targets.cmake` files will take care of the rest. Sadly not many libraries are consumable like this yet, and AFAICT this isn't something vcpkg wants to tackle for its ports (and thus ports have wildly differing ways on how they have to be consumed).

**Comment by MarkSchuckParadigm** [comment]
> And what happens to cmake find modules? They won't find their INCLUDE_DIRs anymore since the don't expect another sub directory.

I'll be honest your comment is over my head. most of my use as a consumer of vcpkg managed dependencies has been find_package(). Looking into the include directory i can see every dependency has its includes thrown together, and as I've explained it causes all sorts of odd problems. From my perspective, i shouldn't be able to create a dependency on something if i didn't' explicitly request it in my cmakelists.

AFAICS find_package() for most of the libraries i use could still work with that use case. As always things are more complicated than a naïve inspection reveals. if its not possible to do anything about it then so be it. This has been a major pain point for me recently, I sure hope it could be fixed.

**Comment by Neumann-A** [comment]
>Sadly not many libraries are consumable like this yet, and AFAICT this isn't something vcpkg wants to tackle for its ports (and thus ports have wildly differing ways on how they have to be consumed).

Simply said. Needs upstream changes/fixes ;). Not something VCPKG should care about. 

>MHO (vcpkg-built) libraries should only ever be included using find_package(<name> CONFIG REQUIRED)

Not all libraries are built with cmake and probably will never be. Furthermore supplying unofficial targets (e.g. like Conan) makes downstream packages not agnostic to the package manager which is bad design.

**Comment by gix** [comment]
> Not all libraries are built with cmake and probably will never be.

That doesn't really matter. What matters is that the library is *consumed* with cmake.

> Furthermore supplying unofficial targets (e.g. like Conan) makes downstream packages not agnostic to the package manager which is bad design.

That's not true. `find_package` has nothing to do with `vcpkg` or any other package manager. You could do this manually if you wanted (but this is cumbersome, and something I'd expect a package manager to take care off). For the ports I use I've added targets files where missing. Makes it really easy to integrate a library and beats the manual `find_include`, `find_library`, `add imported target` dance.

Also, `vcpkg` already renames library names (like adding a `d` suffix) and thus introduces incompatibility.

**Comment by strega-nil** [comment]
This is one of those things that sounds great in theory, but is almost impossible to do in practice. In a perfect world, yes this is how it would work. I'd personally love this at least.

**Comment by MarkSchuckParadigm** [comment]
its been a while since I first posted. I think the main motivation has been solved with more recent VCPKG versions:

use vcpkg.json and its ability to specify dependencies manually per project
only the specified dependencies get installed to 'cmake_outputdir/<triplet>/include/'.

I've switched my projects over to this method on mac/windows/linux and it is working great so far and i can confirm that i can't accidentally include a header that is part of a project that i didn't previously specify in vcpkg.json. So although i can still get link errors if i forgot to add a target_link_libraries to one of the dependencies, at least i can be confident the dependency was listed in vcpkg.json.


VCPKG keeps getting better, Thanks!
